pkg/controller/certificatesigningrequests/vault/vault.go

```diff
@@ -0,0 +1,159 @@
+// provided Vault Issuer or ClusterIssuer. This function updates the
+// CertificateSigningRequest resource if signing was successful. Returns an
+// error which, if not nil, should trigger a retry.
+func (v *Vault) Sign(ctx context.Context, csr *certificatesv1.CertificateSigningRequest, issuerObj cmapi.GenericIssuer) error {
+	log := logf.FromContext(ctx, "sign")
+	log = logf.WithRelatedResource(log, issuerObj)
+
+	resourceNamespace := v.issuerOptions.ResourceNamespace(issuerObj)
+
+	client, err := v.clientBuilder(resourceNamespace, v.secretsLister, issuerObj)
+	if apierrors.IsNotFound(err) {
+		message := "Required secret resource not found"
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "SecretNotFound", message)
+		util.CertificateSigningRequestSetFailed(csr, "SecretNotFound", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	if err != nil {
+		message := fmt.Sprintf("Failed to initialise vault client for signing: %s", err)
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorVaultInit", message)
+		return err
+	}
+
+	duration, err := pki.DurationFromCertificateSigningRequest(csr)
+	if err != nil {
+		message := fmt.Sprintf("Failed to parse requested duration: %s", err)
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorParseDuration", message)
+		util.CertificateSigningRequestSetFailed(csr, "ErrorParseDuration", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	certPEM, caPEM, err := client.Sign(csr.Spec.Request, duration)
+	if err != nil {
+		message := fmt.Sprintf("Vault failed to sign: %s", err)
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorSigning", message)
+		util.CertificateSigningRequestSetFailed(csr, "ErrorSigning", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	log.V(logf.DebugLevel).Info("certificate issued")
+
+	// Kubernetes sub-resources, namely 'status', are separate API endpoints.
+	// We don't want to fire another re-sync of this CertificateSigningRequest
+	// before the `status.Certificate` field has been set as this will fire
+	// another sign call.
+	// Update the status.certificate first so that the sync from updating will
+	// not cause another issuance before setting the CA.
+	csr.Status.Certificate = certPEM
+	csr, err = v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+	if err != nil {
+		message := "Error updating certificate"
+		v.recorder.Eventf(csr, corev1.EventTypeWarning, "ErrorUpdate", "%s: %s", message, err)
+		return err
+	}
+
+	if csr.Annotations == nil {
+		csr.Annotations = make(map[string]string)
+	}
+	csr.Annotations[experimentalapi.CertificateSigningRequestCAAnnotationKey] = base64.StdEncoding.EncodeToString(caPEM)
+	_, err = v.certClient.Update(ctx, csr, metav1.UpdateOptions{})
```

Comment:
context: we discussed this in person, and the above does seem to be the case, that the Certificate field being set is the terminal state for a cert. this might make the case for removing this annotation - if we don't need it.

Reply:
I just looked at [this](https://github.com/jetstack/cert-manager/blob/15b6f32071b45687b44eb097145815fc3af1b010/pkg/controller/certificatesigningrequests/sync.go#L61-L64):

```go
if len(csr.Status.Certificate) > 0 {
	dbg.Info("certificate field is already set in status so skipping processing")
	return nil
}
```

We seem to be using the state of the certificate to branch the logic of the controller itself. This seems like a bad idea, since it means the status is now both consumed and produced by the controller itself.

Reply:
After looking at it again, I realize that we can't "look at the state of the world" with the signed certificate. We have no way to know whether or not we have already signed the request previously.

I think the only way to keep the state consistent is to finish with the `UpdateStatus` even though having first the `Update` and then the `UpdateStatus` may induce two consecutive Sign calls in case the `Update` fails.

Reply:
Argh, nope, my idea can't work...

To summarize, currently, the steps are:

1.  Sign the certificate,
2.  `UpdateStatus` the CSR's status with the signed certificate,
3.  `Update` the CSR with a new annotation that contains the CA.

which may lead to an inconsistency (a CSR has a certificate in its status but no annotation). A consistent order could be:

1.  Sign the certificate,
2.  `Update` the CSR with a new annotation that contains the CA,
3.  `UpdateStatus` the CSR's status with the signed certificate.

This time, no more inconsistencies. The only caveat is that we may sign twice in a row if the first `Update` fails, but this is totally bearable... Except no! It cannot work because the first `Update` will trigger an immediate resync in which the `Sign` function will be re-called, creating an infinite loop.

As Josh detailed in https://github.com/jetstack/cert-manager/pull/4143, the solution is to avoid setting data to both the resource and the status subresource.